### PR TITLE
[MRG] FIX LMNN gradient and cost function

### LINKED ISSUE #46: LMNN - objective calculate?
The current implementation of [LMNN ](https://github.com/all-umass/metric-learn/blob/master/metric_learn/lmnn.py)python version.
I read the paper of LMNN of NIPS05 version, jmlr08 version, I find that the objective of the optimization process in LMNN consisting of the hinge loss of impostors, namely
`$$[1+||L(x_i - x_j)||^2-||L(x_i - x_l)||^2]_+$$`

But the current implementation is
```python
objective = total_active * (1-reg)
objective += G.flatten().dot(L.T.dot(L).flatten())
```
I guess the 2-nd row is calculating the trace of  M * C, but I cannot figure where is the hinge loss?

### PR DESCRIPTION
**Pull request by wdevazelhes**
Fixes #46 
There seems to be problems in LMNN, that we might want to fix for the next release: 

(Note: to find out what to do we can also compare to https://github.com/scikit-learn/scikit-learn/pull/8602 )

-  Looking into this line the gradient seems to be of the shape of an outer product of Xs (when it should be of the shape of L) (see https://github.com/metric-learn/metric-learn/pull/193#issuecomment-489228612). (There might be a dot product with L missing:
(cf. https://github.com/scikit-learn/scikit-learn/pull/8602/files#diff2914dd42441ed8b594e69b2fe04d23f5R762)) This blocks PR #193:

https://github.com/metric-learn/metric-learn/blob/d4badc8adc0a8ce6689dd9a95e89181efaf5ee24/metric_learn/lmnn.py#L196

- Some part of the loss function seems to be an integer count, when it should be the hinge loss ? as suggested by #46: 

https://github.com/metric-learn/metric-learn/blob/d4badc8adc0a8ce6689dd9a95e89181efaf5ee24/metric_learn/lmnn.py#L198

I created a test that should pass when this is fixed (like the one in scikit-learn's PR)

I'll look into it, but I'm not yet that familiar with LMNN so you might know how to fix it

### CODE REVIEW

**Comment by bellet** [comment]
You must be careful when comparing with https://github.com/scikit-learn/scikit-learn/pull/8602 because this version uses a *squared* hinge loss, which obviously changes the cost function and its gradient

**Comment by wdevazelhes** [comment]
I realized in fact there was just one matrix product with L that was missing, I added it and added a toy example too for the loss function

File: metric_learn/lmnn.py [hunk 1]
```diff
@@ -191,10 +191,12 @@ def _loss_grad(self, X, L, dfG, impostors, it, k, reg, target_neighbors, df,
     # do the gradient update
     assert not np.isnan(df).any()
     G = dfG * reg + df * (1 - reg)
+
+    grad = 2 * L.dot(G)
```
Comment (wdevazelhes) [inline metric_learn/lmnn.py]: I put the multiplication by 2 here so that we can test the real gradient using `check_grad`

File: metric_learn/lmnn.py [hunk 2]
```diff
@@ -191,10 +191,12 @@ def _loss_grad(self, X, L, dfG, impostors, it, k, reg, target_neighbors, df,
...
     # compute the objective function
     objective = total_active * (1 - reg)
```
Comment (wdevazelhes) [inline metric_learn/lmnn.py]: In fact this was right, it's the term coming from the +1 that is in the hinge loss (cf. question from #46)

**Comment by wdevazelhes** [comment]
I guess it is ready to merge now

File: test/metric_learn_test.py [hunk 3]
```diff
@@ -120,6 +120,112 @@ def test_iris(self):
...
+        return lmnn._loss_grad(X, L.reshape(-1, X.shape[1]), dfG, impostors,
+                               1, k, reg, target_neighbors, df.copy(),
+                               list(a1), list(a2))[1]
+
+    def grad(L):
+        # we copy variables that can be modified by _loss_grad, because we
+        # want to have the same result when applying the function twice
+        return lmnn._loss_grad(X, L.reshape(-1, X.shape[1]), dfG, impostors,
+                               1, k, reg, target_neighbors, df.copy(),
+                               list(a1), list(a2))[0].ravel()
```
Comment (perimosocordiae) [inline test/metric_learn_test.py]: This test seems particularly brittle, as it replicates so much LMNN functionality. Ideally we could structure the LMNN code to make this simpler, but at the least we should avoid copy-paste here:

```python
def loss_grad(flat_L):
  return lmnn._loss_grad(X, flat_L.reshape(...), ...)

fun = lambda x: loss_grad(x)[1]
grad = lambda x: loss_grad(x)[0].ravel()
```

File: test/metric_learn_test.py [hunk 4]
```diff
@@ -120,6 +120,112 @@ def test_iris(self):
...
+                np.linalg.norm(approx_fprime(L.ravel(), fun, epsilon)))
+    np.testing.assert_almost_equal(rel_diff, 0., decimal=5)
+
+
+@pytest.mark.parametrize('X, y, loss', [(np.array([[0], [1], [2], [3]]),
+                                         [1, 1, 0, 0], 3.0),
+                                        (np.array([[0], [1], [2], [3]]),
+                                         [1, 0, 0, 1], 26.)])
+def test_toy_ex_lmnn(X, y, loss):
+  """Test that the loss give the right result on a toy example"""
```
Comment (perimosocordiae) [inline test/metric_learn_test.py]: Where did you get these values from?

Comment (wdevazelhes) [reply test/metric_learn_test.py]: I computed them by hand, according to the paper (note that it's only the loss function and that I set L to identity so it's easy since the distances are easy to get from this 1D example).
Maybe I can add a small detail of the computation as a comment ?

**Comment by wdevazelhes** [comment]
> Empirically, how does this PR affect the behavior of LMNN on iris / make_blobs / sandwich / etc ?

This is the difference between master and this PR for LMNN: 

master: 
image

this PR: 
image

So I guess a significant improvement, I'll check for the other datasets

**Comment by wdevazelhes** [comment]
But I agree I haven't been through every computation of the algorithm, I just looked more or less if the loss is correct at initialization (it seems to be on the toy examples), but I didn't check the part on recomputing which constraints are active after each iteration etc
But if we assume the loss is correct, it is, from the code:
`loss = constant + tr(A L.T L)`, with `A = dfG * reg + df * (1 - reg) = constant`, 
so the gradient is necessarily `grad = L (A + A.T) = 2 * L A` ( deriving from equation 3 page of this course: http://cs229.stanford.edu/notes/cs229-notes1.pdf)
So I guess now there is just need to check whether the matrix A is the right one, (i.e. it contains the right outer product differences) ?

**Comment by wdevazelhes** [comment]
For iris, we don't see a big difference:

```python
from sklearn.datasets import load_iris
from metric_learn import LMNN
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt
from sklearn.pipeline import make_pipeline

X, y = load_iris(return_X_y=True)

red_dim = make_pipeline(LMNN(), TSNE())
X_e = red_dim.fit_transform(X, y)

plt.scatter(X_e[:, 0], X_e[:, 1], c=y)
plt.show()
```

With master: 
image

With this PR: 
image

**Comment by wdevazelhes** [comment]
From this example (inspired from #180, i.e. a dataset with two informative features and a big noise for the other features), the difference is pretty clear: 
```python
from sklearn.utils import shuffle
from sklearn.datasets import make_classification
from metric_learn import LMNN
from sklearn.manifold import TSNE
import matplotlib.pyplot as plt
from sklearn.pipeline import make_pipeline

X, y = make_classification(random_state=42, n_redundant=0, shuffle=False)
X[:, 2:] *= 10
X, y = shuffle(X, y)

red_dim = make_pipeline(LMNN(), TSNE())
X_e = red_dim.fit_transform(X, y)

plt.scatter(X_e[:, 0], X_e[:, 1], c=y)
plt.show()
```
With this PR: 
image

With master: 
image

File: test/metric_learn_test.py (hunk 3 above)
Comment (wdevazelhes) [reply test/metric_learn_test.py]: Yes you're right, let's go for this clearer expression
done
(note: I didn't use the lambda expression since I was returned a flake8 error, but maybe it's something we want to ignore and still go for the lambda ?)

**Comment by wdevazelhes** [comment]
Sorry my last commit was for the wrong branch, I'll revert it

**Comment by bellet** [comment]
> > Empirically, how does this PR affect the behavior of LMNN on iris / make_blobs / sandwich / etc ?
> 
> This is the difference between master and this PR for LMNN:
> 
> master:
> image
> 
> this PR:
> image
> 
> So I guess a significant improvement, I'll check for the other datasets

This makes me think that we should fix a random seed in the sandwich example so that it is stable when run several times ?

**Comment by bellet** [comment]
What about the problem in the objective function?

File: metric_learn/lmnn.py [hunk 5]
```diff
@@ -191,10 +191,11 @@ def _loss_grad(self, X, L, dfG, impostors, it, k, reg, target_neighbors, df,
     # do the gradient update
     assert not np.isnan(df).any()
     G = dfG * reg + df * (1 - reg)
+    G = L.dot(G)
     # compute the objective function
     objective = total_active * (1 - reg)
-    objective += G.flatten().dot(L.T.dot(L).flatten())
-    return G, objective, total_active, df, a1, a2
+    objective += G.flatten().dot(L.flatten())
+    return 2 * G, objective, total_active, df, a1, a2
```
Comment (bellet) [inline metric_learn/lmnn.py]: I guess the factor 2 comes from differentiating `L.T.dot(L)`. It would be clearer to put it directly in `G` then

**Comment by wdevazelhes** [comment]
> What about the problem in the objective function?

You mean the one with the count being added ? I think it's maybe normal in the end (see comment https://github.com/metric-learn/metric-learn/pull/201#discussion_r286015109). In fact this count should be the sum of all the "+ 1" (the margin) in the hinge losses that are not 0 in the total sum. (But I didn't check yet whether `total_active` is indeed this number of active hinge losses)

**Comment by bellet** [comment]
> > What about the problem in the objective function?
> 
> You mean the one with the count being added ? I think it's maybe normal in the end (see comment [#201 (comment)](https://github.com/metric-learn/metric-learn/pull/201#discussion_r286015109)). In fact this count should be the sum of all the "+ 1" (the margin) in the hinge losses that are not 0 in the total sum. (But I didn't check yet whether `total_active` is indeed this number of active hinge losses)

The hinge loss is equal to max(0, 1 - a). (in the case of LMNN, a is the distance with the impostor minus the distance with the target neighbor)
It is equal to 0 when the margin is satisfied (a >= 1). if not it is equal to the margin violation (1 - a), not "+1"...

**Comment by wdevazelhes** [comment]
> > 
> 
> The hinge loss is equal to max(0, 1 - a). (in the case of LMNN, a is the distance with the impostor minus the distance with the target neighbor)
> It is equal to 0 when the margin is satisfied (a >= 1). if not it is equal to the margin violation (1 - a), not "+1"...

I was not thinking that "+1'" only was the value of the hinge loss, but I was thinking it was the "+1" part of the hinge loss (sum(1-a) = sum(1) - sum(a) = **total_active** - sum(a)). I had the impression that the "df" part later (some outer products), with operations with L, accounted for the sum of "a"s (distance with the impostor minus the distance with the target neighbor) so in total there was something like: `push_loss = total_active - sum_of_the_differences_btw_distance_with_the_impostor_and_distance _with_target_neighbor`
But I definitely need to check if this is indeed the case I didn't deepen the "df" computation part

> You can run some executions showing how the objective function evolves to make sure it goes down at each iteration when small enough step size is used, and compare with the version in master (the number of active constraints should tend to go down but not each time)

Yes good idea I'll do that

**Comment by wdevazelhes** [comment]
Oups, in fact I introduced this bug in PR #101: I forgot to multiply by L when applying the update...

**Comment by bellet** [comment]
Ok fair enough - the objective may be correct but the code is very hard to parse. Let's fix the gradient issue for now and make an issue to maybe refactor the code later to make it more easy to maintain.

**Comment by wdevazelhes** [comment]
So in the end this PR just fixes the bug that I introduced, as well as introduced some tests, so if everybody agrees, I think we are good to merge ? @perimosocordiae ?

**Comment by perimosocordiae** [comment]
Merged!
